### PR TITLE
Overhaul DataFrame plannning to use DataFusion Extensions.

### PR DESCRIPTION
**Pull request by jacksonrnewhouse**
This gets rid of the intermediate LogicalPlanExtension graph in favor of rewriting the logical plan, making more use of DataFusion's Extensions. In particular, I've added extensions for aggregations, joins, key calculations, and sources. There are also more synthetic extensions that are there to clarify the physical boundaries and simulate record batch changes that will be executed by the operators. For instance, the tumbling aggregating operator manually inserts the timestamp column after aggregations are calculated, and this is represented by the TimestampAppendExtension.

### CODE REVIEW

File: crates/arroyo-df/src/direct.rs
Comment (mwylde) [inline crates/arroyo-df/src/direct.rs:1]: it's not obvious to me what "direct" means. Is there a more clear name for this file?

File: crates/arroyo-df/src/rewriters.rs [hunk 1]
```diff
@@ -29,10 +33,17 @@ impl TreeNodeRewriter for TimestampRewriter {
         match node {
             LogicalPlan::Projection(ref mut projection) => {
                 if !has_timestamp_field(projection.schema.clone()) {
-                    projection.schema = add_timestamp_field(projection.schema.clone(), None)
-                        .expect("in projection");
+                    let timestamp_field = projection
```
Comment (mwylde) [inline crates/arroyo-df/src/rewriters.rs:36]: It looks like TimestampRewriter is no longer used, can we remove it?

**Review by mwylde — COMMENTED** [review]
* This query fails:

```
create table impulse with (
    connector = 'impulse',
    event_rate = '5'
);
 
SELECT avg(counter)
from impulse
group by hop('1 second', '5 seconds');
```

with 

```
2024-02-27T19:19:45.792874Z ERROR arroyo_server_common: panicked at /Users/mwylde/git/arroyo/crates/arroyo-operator/src/context.rs:261:22:
called `Result::unwrap()` on an `Err` value: SendError { .. } panic.file="/Users/mwylde/git/arroyo/crates/arroyo-operator/src/context.rs" panic.line=261 panic.column=22
```

* UDFs don't seem to work. I get this error:

```
2024-02-27T19:21:11.794799Z ERROR arroyo_server_common: panicked at crates/arroyo-worker/src/engine.rs:804:10:
Failed to construct operator ArrowValue: Execution error: Udf new_udf not found panic.file="crates/arroyo-worker/src/engine.rs" panic.line=804 panic.column=10
```

File: crates/arroyo-df/src/direct.rs [hunk 2]
```diff
@@ -0,0 +1,888 @@
...
+        if let Some(table_index) = self.named_nodes.get(&table_source.name) {
+            self.add_index_to_traversal(*table_index);
+            return Ok(());
+        }
+
+        let sql_source = table_source
+            .table
+            .as_sql_source()
+            .map_err(|e| DataFusionError::Plan(format!("as_sql_source error: {}", e)))?;
+        let node = LogicalNode {
```
[the remark this answers is not in the document]
Comment (jacksonrnewhouse) [reply crates/arroyo-df/src/direct.rs:152]: This would indicate a bug, I believe, but I'll clean it up.

**Review by mwylde — APPROVED** [review]

**Pull request merged**